WPMedia/fusion-news-theme-blocks
# PEN-1796: Lazy Load Blocks (based on RC)

**If you have not filled out the checklist below, the pr is not ready for review.**

## Description
Implemented <LazyLoad/> SDK component in several blocks, modified/fixed unit tests

## Jira Ticket
- [PEN-1796](https://arcpublishing.atlassian.net/browse/PEN-1796)

## Acceptance Criteria
![image](https://user-images.githubusercontent.com/26662906/109243708-04de8a80-77a3-11eb-8ac9-2530132053eb.png)

## Test Steps
1. In `Fusion-News-Theme` repo, checkout `master` & `git pull`
2. After checking out this feature branch in `fusion-news-theme-blocks`, run `npx lerna clean -y && rm -rf node_modules && npm i && npx lerna clean -y`
3. In your local `engine-theme-sdk` repo, checkout this branch https://github.com/WPMedia/fusion-news-theme-blocks/pull/755 and then run `rm -rf ./node_modules && npm i && npm run build`
4. In your `Fusion-News-Theme``.env`, make sure your `ENGINE_SDK_REPO` variable is set to the location of your engine SDK repo
5. Run `npx fusion start -f -l @wpmedia/ads-block,@wpmedia/header-nav-chain-block,@wpmedia/article-body-block,@wpmedia/author-bio-block,@wpmedia/footer-block,@wpmedia/gallery-block,@wpmedia/share-bar-block,@wpmedia/small-manual-promo-block,@wpmedia/medium-manual-promo-block,@wpmedia/large-manual-promo-block,@wpmedia/extra-large-manual-promo-block,@wpmedia/small-promo-block,@wpmedia/medium-promo-block,@wpmedia/large-promo-block,@wpmedia/extra-large-promo-block,@wpmedia/tag-title-block,@wpmedia/top-table-list-block,@wpmedia/header-block,@wpmedia/video-player-block,@wpmedia/tag-title-block,@wpmedia/simple-list-block,@wpmedia/numbered-list-block,@wpmedia/results-list-block,@wpmedia/results-list-block,@wpmedia/search-results-list-block,@wpmedia/card-list-block,@wpmedia/article-tag-block` in `Fusion-News-Theme`
6. Open the `all-blocks` page in PageBuilder editor
7. In your PageBuilder editor, set the `Lazy Load block?` checkbox to "checked" for all of the blocks listed in the screenshot below
8. Stage & publish and open http://localhost/all-blocks/?_website=the-gazette in a separate tab
9. You should now see that all of the affected blocks are correctly lazy-loaded. This is a bit difficult to see since it loads quickly as you scroll down the viewport. The best way I've found is to set (in Chrome DevTools) breakpoints and/or log points inside the `Render` component (`CardListRender`, `FooterRender`) that each of the affected blocks now has nested in their component bodies (to wrap the `<LazyLoad/>` component around since it has to be wrapped around a component and not just a block of JSX). You can also _sort of_ tell that it's working if you scroll down the page really fast or use the PgUp/PgDown keys. You can also monitor changes in the DOM through the `Elements` tab but I found this harder to track.

![image](https://user-images.githubusercontent.com/26662906/109243746-1889f100-77a3-11eb-8736-552b6b217134.png)

## Effect Of Changes
### Before
Lazy-loading option was only available on Ads Block previously

### After
Lazy-loading option is now available on all blocks listed in the above screenshot

## Dependencies or Side Effects
_Examples of dependencies or side effects are:_
- Additional settings that will be required in the blocks.json
- Changes to the custom fields which will require users to reconfigure features
- Update to css framework or SDK
- Dependency on another PR that needs to be merged first

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
please explain why (so that we can fix it whenever it gets refactored).

## Reviewer Checklist 
_The reviewer of the PR should copy-paste this template into the review comments on review._
